lib/tc.py

```diff
@@ -0,0 +1,117 @@
+
+        :param flavor: Type of test to run (functional or update).
+        :param properties: Task properties for template rendering
+        """
+        template_file = os.path.join(os.path.dirname(__file__),
+                                     'tasks', '{}.yml'.format(flavor))
+        if not os.path.isfile(template_file):
+            raise errors.NotSupportedException('Test type "{}" not supported.'.format(flavor))
+
+        with open(template_file) as f:
+            template = jinja2.Template(f.read(), undefined=jinja2.StrictUndefined)
+
+        template_vars = copy.deepcopy(properties)
+        template_vars.update({
+            'stableSlugId': taskcluster.stableSlugId(),
+            'now': taskcluster.stringDate(datetime.datetime.utcnow()),
+            'fromNow': taskcluster.fromNow,
+            'docker_task_id': self.get_docker_task_id('desktop-test', properties),
+        })
+
+        rendered = template.render(**template_vars)
+
+        return yaml.safe_load(rendered)
+
+    def get_docker_task_id(self, image_name, properties):
+        """Retrieve docker image task Id used by tests on TC.
+
+        To get the task id of the 'desktop-test' docker image as used by desktop tests
+        for a given build the Index has to be queried first. When the task for the build
+        has been found, check all its dependent tasks for the first appearance of the
+        'desktop-test' worker type. Within its payload the task id of the docker task
+        can be extracted.
+
+        Bug 1276352 - Not all Taskcluster builds are Tier-1 yet. So the index also
+        contains BuildBot entries. To ensure we get a TC build force to Linux64 dbg.
+
+        :param image_name: Name of the docker image to look for.
+        :param properties: Properties of the build and necessary resources.
+        """
+        build_index = 'gecko.v2.{branch}.revision.{rev}.firefox.{platform}-dbg'.format(
+            branch=properties['branch'],
+            rev=properties['revision'],
+            platform=properties['platform'],
+        )
+
+        try:
+            logger.debug('Querying Taskcluster for "desktop-test" docker image for "{}"...'.format(
+                properties['branch']))
+            build_task_id = taskcluster.Index().findTask(build_index)['taskId']
+        except taskcluster.exceptions.TaskclusterFailure:
+            raise errors.NotFoundException('Required build not found for TC index', build_index)
+
+        task_id = None
+        continuation_token = None
+        while not task_id:
+            options = {'limit': 5}
+            if continuation_token:
+                options.update({'continuationToken': continuation_token})
+
+            resp = taskcluster.Queue().listDependentTasks(build_task_id,
+                                                          options=options)
+            for task in resp['tasks']:
+                if task['status']['workerType'] == image_name:
```

Comment:
This assumes that the workerType and the image name match, which happens to be true right now, but is basically a coincidence.  Joel is in the process of changing that, in fact, as he moves tests to larger EC2 instances.  It might be better to use some other way of recognizing a test task here.  Perhaps something like `suite` in `task.extra`?